File: sdk/src/main/java/ru/niisokb/safesdk/java/SpLocationManager.java
package ru.niisokb.safesdk.java;

import android.app.PendingIntent;
import android.content.Context;
import android.location.Criteria;
import android.location.GnssMeasurementsEvent;
import android.location.GnssNavigationMessage;
import android.location.GnssStatus;
import android.location.GpsStatus;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationProvider;
import android.location.OnNmeaMessageListener;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;

import java.util.List;

import androidx.annotation.RequiresApi;

/**
 * LocationManager который учитывает настройки ограничений GPS. Перед использованием необходимо
 * инициализировать класс при помощи [load]. А также для применения конфигураций должен быть
 * загружен SpConfigDispatcher
 */
@SuppressWarnings({"AccessStaticViaInstance", "unused"})
public class SpLocationManager {
    public static String GPS_RESTRICTION_KEY = "safephone.location.gps_enabled";

    private static ru.niisokb.safesdk.SpLocationManager instance =
            ru.niisokb.safesdk.SpLocationManager.INSTANCE;

    public static String NETWORK_PROVIDER = instance.NETWORK_PROVIDER;
    public static String GPS_PROVIDER = instance.GPS_PROVIDER;
    public static String PASSIVE_PROVIDER = instance.PASSIVE_PROVIDER;
    public static String KEY_PROXIMITY_ENTERING = instance.KEY_PROXIMITY_ENTERING;
    public static String KEY_STATUS_CHANGED = instance.KEY_STATUS_CHANGED;
    public static String KEY_PROVIDER_ENABLED = instance.KEY_PROVIDER_ENABLED;
    public static String KEY_LOCATION_CHANGED = instance.KEY_LOCATION_CHANGED;
    public static String PROVIDERS_CHANGED_ACTION = instance.PROVIDERS_CHANGED_ACTION;
    public static String MODE_CHANGED_ACTION = instance.MODE_CHANGED_ACTION;

    public SpLocationManager load(Context context) {
        instance.load(context);
        return this;
    }

    public boolean isInitialized() {
        return instance.isInitialized();
    }

    public boolean isGpsAccessEnabled() {
        return instance.isGpsAccessEnabled();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public boolean addNmeaListener(OnNmeaMessageListener listener, Handler handler) {
        return instance.addNmeaListener(listener, handler);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public boolean addNmeaListener(OnNmeaMessageListener listener) {
        return instance.addNmeaListener(listener);
    }

    public static void addProximityAlert(double latitude, double longitude,
                                         float radius, long expiration, PendingIntent intent) {
        instance.addProximityAlert(latitude, longitude, radius, expiration, intent);
    }

    public static void addTestProvider(String name, boolean requiresNetwork,
                                       boolean requiresSatellite, boolean requiresCell,
                                       boolean hasMonetaryCost, boolean supportsAltitude,
                                       boolean supportsSpeed, boolean supportsBearing,
                                       int powerRequirement, int accuracy) {
        instance.addTestProvider(name, requiresNetwork,
                requiresSatellite, requiresCell,
                hasMonetaryCost, supportsAltitude,
                supportsSpeed, supportsBearing, powerRequirement,
                accuracy);
    }

    public static void clearTestProviderEnabled(String provider) {
        instance.clearTestProviderEnabled(provider);
    }

    public static void clearTestProviderLocation(String provider) {
        instance.clearTestProviderLocation(provider);
    }

    public static void clearTestProviderStatus(String provider) {
        instance.clearTestProviderStatus(provider);
    }

    public static List<String> getAllProviders() {
        return instance.getAllProviders();
    }

    @RequiresApi(api = Build.VERSION_CODES.P)
    public static String getGnssHardwareModelName() {
        return instance.getGnssHardwareModelName();
    }

    @RequiresApi(api = Build.VERSION_CODES.P)
    public static int getGnssYearOfHardware() {
        return instance.getGnssYearOfHardware();
    }

    public static GpsStatus getGpsStatus(GpsStatus status) {
        return instance.getGpsStatus(status);
    }

    public static Location getLastKnownLocation(String provider) {
        return instance.getLastKnownLocation(provider);
    }

    public static LocationProvider getProvider(String name) {
        return instance.getProvider(name);
    }

    public static List<String> getProviders(boolean enabledOnly) {
        return instance.getProviders(enabledOnly);
    }

    public static List<String> getProviders(Criteria criteria, boolean enabledOnly) {
        return instance.getProviders(criteria, enabledOnly);
    }

    @RequiresApi(api = Build.VERSION_CODES.P)
    public static boolean isLocationEnabled() {
        return instance.isLocationEnabled();
    }

    public static boolean isProviderEnabled(String provider) {
        return instance.isProviderEnabled(provider);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static boolean registerGnssMeasurementsCallback(GnssMeasurementsEvent.Callback callback) {
        return instance.registerGnssMeasurementsCallback(callback);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static boolean registerGnssMeasurementsCallback(GnssMeasurementsEvent.Callback callback, Handler handler) {
        return instance.registerGnssMeasurementsCallback(callback, handler);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static boolean registerGnssNavigationMessageCallback(GnssNavigationMessage.Callback callback, Handler handler) {
        return instance.registerGnssNavigationMessageCallback(callback, handler);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static boolean registerGnssNavigationMessageCallback(GnssNavigationMessage.Callback callback) {
        return instance.registerGnssNavigationMessageCallback(callback);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static boolean registerGnssStatusCallback(GnssStatus.Callback callback) {
        return instance.registerGnssStatusCallback(callback);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static boolean registerGnssStatusCallback(GnssStatus.Callback callback, Handler handler) {
        return instance.registerGnssStatusCallback(callback, handler);
    }

    public static void removeGpsStatusListener(GpsStatus.Listener listener) {
        instance.removeGpsStatusListener(listener);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static void removeNmeaListener(OnNmeaMessageListener listener) {
        instance.removeNmeaListener(listener);
    }

    public static void removeProximityAlert(PendingIntent intent) {
        instance.removeProximityAlert(intent);
    }

    public static void removeTestProvider(String provider) {
        instance.removeTestProvider(provider);
    }

    public static void removeUpdates(LocationListener listener) {
        instance.removeUpdates(listener);
    }

    public static void removeUpdates(PendingIntent intent) {
        instance.removeUpdates(intent);
    }

    public static void requestLocationUpdates(String provider, long minTime,
                                              float minDistance, LocationListener listener) {
        instance.requestLocationUpdates(provider, minTime, minDistance, listener);
    }

    public static void requestLocationUpdates(String provider, long minTime, float minDistance,
                                              LocationListener listener, Looper looper) {
        instance.requestLocationUpdates(provider, minTime, minDistance, listener, looper);
    }

    public static void requestLocationUpdates(long minTime, float minDistance, Criteria criteria,
                                              PendingIntent intent) {
        instance.requestLocationUpdates(minTime, minDistance, criteria, intent);
    }

    public static void requestLocationUpdates(String provider, long minTime, float minDistance,
                                              PendingIntent intent) {
        instance.requestLocationUpdates(provider, minTime, minDistance, intent);
    }

    public static void requestSingleUpdate(String provider, PendingIntent intent) {
        instance.requestSingleUpdate(provider, intent);
    }

    public static void requestSingleUpdate(Criteria criteria, PendingIntent intent) {
        instance.requestSingleUpdate(criteria, intent);
    }

    public static void requestSingleUpdate(String provider, LocationListener listener, Looper looper) {
        instance.requestSingleUpdate(provider, listener, looper);
    }

    public static void requestSingleUpdate(Criteria criteria, LocationListener listener, Looper looper) {
        instance.requestSingleUpdate(criteria, listener, looper);
    }

    public static void setTestProviderEnabled(String provider, boolean enabled) {
        instance.setTestProviderEnabled(provider, enabled);
    }

    public static void setTestProviderLocation(String provider, Location loc) {
        instance.setTestProviderLocation(provider, loc);
    }

    public static void setTestProviderStatus(String provider, int status, Bundle extras, long updateTime) {
        instance.setTestProviderStatus(provider, status, extras, updateTime);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static void unregisterGnssMeasurementsCallback(GnssMeasurementsEvent.Callback callback) {
        instance.unregisterGnssMeasurementsCallback(callback);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static void unregisterGnssNavigationMessageCallback(GnssNavigationMessage.Callback callback) {
        instance.unregisterGnssNavigationMessageCallback(callback);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    public static void unregisterGnssStatusCallback(GnssStatus.Callback callback) {
        instance.unregisterGnssStatusCallback(callback);
    }
}

File: sdk/src/main/java/ru/niisokb/safesdk/SpLocationManager.kt
@file:Suppress("unused", "DEPRECATION")

package ru.niisokb.safesdk

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.content.Context
import android.location.*
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.RequiresApi
import ru.niisokb.safesdk.SpLocationManager.load
import ru.niisokb.safesdk.configuration.ConfigAttr
import ru.niisokb.safesdk.mixins.ConfigurableModule
import ru.niisokb.safesdk.configuration.ValType

/**
 * LocationManager, который учитывает настройки ограничений GPS. Перед использованием необходимо
 * инициализировать класс при помощи [load].
 */
@Suppress("MemberVisibilityCanBePrivate")
object SpLocationManager : ConfigurableModule {

    private const val TAG = "SpLocationManager"
    private const val missingMethodExceptionMessage =
            "This method is not available, try to update library to the latest version."

    const val GPS_ENABLED_KEY = "safephone.restrictions.gps_enabled"
    const val GPS_ENABLED_DEFAULT = true

    override val attrs = listOf(ConfigAttr(GPS_ENABLED_KEY, ValType.Boolean, GPS_ENABLED_DEFAULT))
    override val dispatch: (Bundle) -> Unit = { config ->
        gpsEnabled = config.getBoolean(GPS_ENABLED_KEY, GPS_ENABLED_DEFAULT)
    }

    init {
        register()
    }

    private var gpsEnabled = GPS_ENABLED_DEFAULT
    private var initialized = false
    private lateinit var locationManager: LocationManager

    private val gpsDisabledException =
            SpLocationManagerException.GpsDisabled("Gps disabled.")
    private val notInitializedException =
            SpLocationManagerException.NotInitialized("Location manager is not initialized.")


    /** @see LocationManager.NETWORK_PROVIDER */
    const val NETWORK_PROVIDER = "network"

    /** @see LocationManager.GPS_PROVIDER */
    const val GPS_PROVIDER = "gps"

    /** @see LocationManager.PASSIVE_PROVIDER */
    const val PASSIVE_PROVIDER = "passive"

    /** @see LocationManager.KEY_PROXIMITY_ENTERING */
    const val KEY_PROXIMITY_ENTERING = "entering"

    /** @see LocationManager.KEY_STATUS_CHANGED */
    const val KEY_STATUS_CHANGED = "status"

    /** @see LocationManager.KEY_PROVIDER_ENABLED */
    const val KEY_PROVIDER_ENABLED = "providerEnabled"

    /** @see LocationManager.KEY_LOCATION_CHANGED */
    const val KEY_LOCATION_CHANGED = "location"

    /** @see LocationManager.PROVIDERS_CHANGED_ACTION */
    const val PROVIDERS_CHANGED_ACTION = "android.location.PROVIDERS_CHANGED"

    /** @see LocationManager.MODE_CHANGED_ACTION */
    const val MODE_CHANGED_ACTION = "android.location.MODE_CHANGED"

    /** Инициализирует [SpLocationManager]. */
    fun load(context: Context): SpLocationManager {
        if (!initialized) {
            locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager

            initialized = true
        }

        return this
    }

    fun isInitialized() = initialized

    fun isGpsAccessEnabled() = gpsEnabled

    /** @see LocationManager.addNmeaListener */
    @RequiresApi(Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    fun addNmeaListener(listener: OnNmeaMessageListener, handler: Handler): Boolean {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }

        return try {
            locationManager.addNmeaListener(listener, handler)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.addNmeaListener */
    @SuppressLint("MissingPermission")
    @RequiresApi(Build.VERSION_CODES.N)
    fun addNmeaListener(listener: OnNmeaMessageListener): Boolean {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }

        return try {
            locationManager.addNmeaListener(listener)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.addProximityAlert */
    @SuppressLint("MissingPermission")
    fun addProximityAlert(latitude: Double, longitude: Double, radius: Float,
                          expiration: Long, intent: PendingIntent) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }

        try {
            locationManager.addProximityAlert(latitude, longitude, radius, expiration, intent)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.addTestProvider */
    fun addTestProvider(name: String, requiresNetwork: Boolean, requiresSatellite: Boolean,
                        requiresCell: Boolean, hasMonetaryCost: Boolean, supportsAltitude: Boolean,
                        supportsSpeed: Boolean, supportsBearing: Boolean, powerRequirement: Int,
                        accuracy: Int) {
        if (!initialized) {
            throw notInitializedException
        }

        try {
            locationManager.addTestProvider(name, requiresNetwork, requiresSatellite,
                    requiresCell, hasMonetaryCost, supportsAltitude,
                    supportsSpeed, supportsBearing, powerRequirement, accuracy)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)

        }
    }

    /** @see LocationManager.clearTestProviderEnabled */
    fun clearTestProviderEnabled(provider: String) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.clearTestProviderEnabled(provider)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)

        }

    }

    /** @see LocationManager.clearTestProviderLocation */
    fun clearTestProviderLocation(provider: String) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.clearTestProviderLocation(provider)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.clearTestProviderStatus */
    fun clearTestProviderStatus(provider: String) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.clearTestProviderLocation(provider)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.getAllProviders */
    fun getAllProviders(): List<String> {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.allProviders
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            ArrayList()
        }
    }

    /** @see LocationManager.getBestProvider */
    fun getBestProvider(criteria: Criteria, enabledOnly: Boolean): String {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.getBestProvider(criteria, enabledOnly)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            ""
        }
    }

    /** @see LocationManager.getGnssHardwareModelName */
    @RequiresApi(Build.VERSION_CODES.P)
    @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
    fun getGnssHardwareModelName(): String {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.gnssHardwareModelName
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            ""
        }

    }

    /** @see LocationManager.getGnssYearOfHardware */
    @RequiresApi(Build.VERSION_CODES.P)
    fun getGnssYearOfHardware(): Int {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.gnssYearOfHardware
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            -1
        }
    }

    /** @see LocationManager.getGpsStatus */
    @SuppressLint("MissingPermission")
    fun getGpsStatus(status: GpsStatus): GpsStatus {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.getGpsStatus(status)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            status
        }
    }

    /** @see LocationManager.getLastKnownLocation */
    @SuppressLint("MissingPermission")
    fun getLastKnownLocation(provider: String): Location {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.getLastKnownLocation(provider)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            Location(provider)
        }
    }

    /** @see LocationManager.getProvider */
    fun getProvider(name: String): LocationProvider? {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.getProvider(name)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            null
        }
    }

    /** @see LocationManager.getProvider */
    fun getProviders(enabledOnly: Boolean): List<String> {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.getProviders(enabledOnly)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            ArrayList()
        }
    }

    /** @see LocationManager.getProviders */
    fun getProviders(criteria: Criteria, enabledOnly: Boolean): List<String> {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.getProviders(criteria, enabledOnly)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            ArrayList()
        }
    }

    /** @see LocationManager.isLocationEnabled */
    @RequiresApi(Build.VERSION_CODES.P)
    fun isLocationEnabled(): Boolean {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.isLocationEnabled
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.isProviderEnabled */
    fun isProviderEnabled(provider: String): Boolean {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.isProviderEnabled(provider)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.registerGnssMeasurementsCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    fun registerGnssMeasurementsCallback(callback: GnssMeasurementsEvent.Callback): Boolean {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }

        return try {
            locationManager.registerGnssMeasurementsCallback(callback)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.registerGnssMeasurementsCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    fun registerGnssMeasurementsCallback(callback: GnssMeasurementsEvent.Callback,
                                         handler: Handler): Boolean {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.registerGnssMeasurementsCallback(callback, handler)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.registerGnssMeasurementsCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    fun registerGnssNavigationMessageCallback(callback: GnssNavigationMessage.Callback,
                                              handler: Handler): Boolean {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.registerGnssNavigationMessageCallback(callback, handler)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.registerGnssNavigationMessageCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    fun registerGnssNavigationMessageCallback(callback: GnssNavigationMessage.Callback): Boolean {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.registerGnssNavigationMessageCallback(callback)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.registerGnssStatusCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    fun registerGnssStatusCallback(callback: GnssStatus.Callback): Boolean {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.registerGnssStatusCallback(callback)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.registerGnssStatusCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    @SuppressLint("MissingPermission")
    fun registerGnssStatusCallback(callback: GnssStatus.Callback, handler: Handler): Boolean {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }

        return try {
            locationManager.registerGnssStatusCallback(callback, handler)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.removeGpsStatusListener */
    fun removeGpsStatusListener(listener: GpsStatus.Listener) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.removeGpsStatusListener(listener)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.removeNmeaListener */
    @RequiresApi(Build.VERSION_CODES.N)
    fun removeNmeaListener(listener: OnNmeaMessageListener) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.removeNmeaListener(listener)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.removeProximityAlert */
    @SuppressLint("MissingPermission")
    fun removeProximityAlert(intent: PendingIntent) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.removeProximityAlert(intent)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.removeTestProvider */
    fun removeTestProvider(provider: String) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.removeTestProvider(provider)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.removeUpdates */
    @SuppressLint("MissingPermission")
    fun removeUpdates(listener: LocationListener) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.removeUpdates(listener)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.removeUpdates */
    fun removeUpdates(intent: PendingIntent) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.removeUpdates(intent)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestLocationUpdates */
    @SuppressLint("MissingPermission")
    fun requestLocationUpdates(provider: String, minTime: Long,
                               minDistance: Float, listener: LocationListener) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestLocationUpdates(provider, minTime, minDistance, listener)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestLocationUpdates */
    @SuppressLint("MissingPermission")
    fun requestLocationUpdates(minTime: Long, minDistance: Float, criteria: Criteria,
                               listener: LocationListener, looper: Looper) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestLocationUpdates(minTime, minDistance,
                    criteria, listener, looper)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestLocationUpdates */
    @SuppressLint("MissingPermission")
    fun requestLocationUpdates(provider: String, minTime: Long, minDistance: Float,
                               listener: LocationListener, looper: Looper) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestLocationUpdates(provider, minTime, minDistance,
                    listener, looper)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestLocationUpdates */
    @SuppressLint("MissingPermission")
    fun requestLocationUpdates(minTime: Long, minDistance: Float,
                               criteria: Criteria, intent: PendingIntent) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestLocationUpdates(minTime, minDistance, criteria, intent)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestLocationUpdates */
    @SuppressLint("MissingPermission")
    fun requestLocationUpdates(provider: String, minTime: Long,
                               minDistance: Float, intent: PendingIntent) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestLocationUpdates(provider, minTime, minDistance, intent)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestSingleUpdate */
    @SuppressLint("MissingPermission")
    fun requestSingleUpdate(provider: String, intent: PendingIntent) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestSingleUpdate(provider, intent)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestSingleUpdate */
    @SuppressLint("MissingPermission")
    fun requestSingleUpdate(criteria: Criteria, intent: PendingIntent) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestSingleUpdate(criteria, intent)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.requestSingleUpdate */
    @SuppressLint("MissingPermission")
    fun requestSingleUpdate(provider: String, listener: LocationListener, looper: Looper) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestSingleUpdate(provider, listener, looper)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.requestSingleUpdate */
    @SuppressLint("MissingPermission")
    fun requestSingleUpdate(criteria: Criteria, listener: LocationListener, looper: Looper) {
        if (!gpsEnabled) {
            throw gpsDisabledException
        }
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.requestSingleUpdate(criteria, listener, looper)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }

    }

    /** @see LocationManager.sendExtraCommand */
    fun sendExtraCommand(provider: String, command: String, extras: Bundle): Boolean {
        if (!initialized) {
            throw notInitializedException
        }
        return try {
            locationManager.sendExtraCommand(provider, command, extras)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
            false
        }
    }

    /** @see LocationManager.setTestProviderEnabled */
    fun setTestProviderEnabled(provider: String, enabled: Boolean) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.setTestProviderEnabled(provider, enabled)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.setTestProviderLocation */
    fun setTestProviderLocation(provider: String, loc: Location) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.setTestProviderLocation(provider, loc)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.setTestProviderStatus */
    fun setTestProviderStatus(provider: String, status: Int, extras: Bundle, updateTime: Long) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.setTestProviderStatus(provider, status, extras, updateTime)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.unregisterGnssMeasurementsCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    fun unregisterGnssMeasurementsCallback(callback: GnssMeasurementsEvent.Callback) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.unregisterGnssMeasurementsCallback(callback)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.unregisterGnssNavigationMessageCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    fun unregisterGnssNavigationMessageCallback(callback: GnssNavigationMessage.Callback) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.unregisterGnssNavigationMessageCallback(callback)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }

    /** @see LocationManager.unregisterGnssStatusCallback */
    @RequiresApi(Build.VERSION_CODES.N)
    fun unregisterGnssStatusCallback(callback: GnssStatus.Callback) {
        if (!initialized) {
            throw notInitializedException
        }
        try {
            locationManager.unregisterGnssStatusCallback(callback)
        } catch (e: NoSuchMethodException) {
            Log.w(TAG, missingMethodExceptionMessage)
        }
    }
}

sealed class SpLocationManagerException(msg: String) : RuntimeException(msg) {
    class GpsDisabled(msg: String) : SpLocationManagerException(msg)
    class NotInitialized(msg: String) : SpLocationManagerException(msg)
}
